### ISSUE TITLE
useDataProvider - dataProvider.create calls buildQuery multiple times without forwarding "resource"

### ISSUE THREAD

**Issue by Vadorequest** [marmelab/react-admin#4546]
I'm trying to build an `AssetInput` component, which uses `filestack-react` library to upload an asset to FIleStack provider, using react-admin. The goal -in this example- is to upload a customer's logo through the Customer Edit view.

The upload itself (through `filestack-react`) works fine, then a GraphQL mutation (CREATE) is executed on my GraphQL server, and it  works too, but the client crashes for a reason I failed to locate.

Basically, there isn't one call made to my data provider `buildQuery` function, but more like 5 (instead of 1). And those call don't forward the `resource` nor `data` properties, which makes the whole thing crash on the client side.

I'm using `useDataProvider` hook.

Here is the source code:

```tsx
/** @jsx jsx */
import { jsx } from '@emotion/core';
import React from 'react';
import { useDataProvider } from 'react-admin';

import { FileStackResult } from '../../../types/fileStack/FileStackResult';
import { FileStackUploadedFile } from '../../../types/fileStack/FileStackUploadedFile';
import { Record } from '../../../utils/record';

const AssetInput = (props: Props) => {
  const {
    label,
    source,
    record,
  } = props;
  const FileStackAssetPicker = require('filestack-react').default;
  console.debug('AssetInput.props', props);
  const dataProvider = useDataProvider();

  return (
    <div>
      {label}
      <FileStackAssetPicker
        apikey={process.env.FILESTACK_API_KEY}
        onSuccess={async (result: FileStackResult) => {
          const { filesUploaded } = result;
          console.log('filesUploaded', filesUploaded);
          const fileUploaded: FileStackUploadedFile = filesUploaded[0];
          console.log('fileUploaded', fileUploaded);

          if (fileUploaded) {
            console.log('dataProvider', dataProvider);
            try {
              const res: any = await dataProvider.create('Asset', {
                data: {
                  handle: fileUploaded.handle,
                  fileName: fileUploaded.filename,
                  title: fileUploaded.filename,
                  size: fileUploaded.size,
                  mimeType: fileUploaded.mimeType || 'image/png', // TODO resolve proper mimeType
                  status: 'PUBLISHED',
                },
              });
              console.debug('res', res);

            } catch (e) {
              // Do nothing. (Doesn't work at all, errors are still thrown and not caught there)
            }
          }
        }}
      />
    </div>
  );
};

type Props = {
  source: string;
  record?: Record;
  label?: string;
}

export default AssetInput;

```

**What you were expecting:**
I'm expecting that `dataProvider.create(resource, params)` would only call once my custom `enhancedBuildQuery`, instead of 5. And that it would forward resource and params correctly.

**What happened instead:**
- **Video** (shows the workflow and the bug, explain the code a bit): https://youtu.be/ZapApLIUrAk (5mn)
- **Live demo**: https://nrn-admin-filestack-upload.now.sh/#/Customer/ck73s1dqm2ue30b98uepy9is3 (display the browser console and try to "Pick file")

**Steps to reproduce:**
I guess it may be related to `useDataProvider`, or `dataProvider.create`. I couldn't define precisely where the issue comes from.

**Related code:**
- PR: https://github.com/UnlyEd/next-right-now-admin/pull/29
- Most interesting commit: https://github.com/UnlyEd/next-right-now-admin/pull/29/commits/9c68e85661b16e83c2494b5bff8e4620c7436e55

**Environment**

* React-admin version: 3.3.0
* React version: 16.12.0

**Comment by Vadorequest** [comment]
Added live demo: https://nrn-admin-filestack-upload.now.sh/#/Customer/ck73s1dqm2ue30b98uepy9is3

**Comment by Vadorequest** [comment]
Using step by step debug into `useDataProvider`, I noticed it's indeed called multiple times with various arguments.

```js
var dataProviderProxy = useMemo(function () {
        return new Proxy(dataProvider, {
            get: function (target, name) {
                return function (resource, payload, options) {
                    var type = name.toString(); // Breakpoint here
```

There are 6 calls, the first one has all parameters `resource, payload, options` set to `undefined`. 
Then, `resource` is set to `string`, but the other are still `undefined`. 
Eventually, one of them is correct with `resource` set to `Asset` as expected, and that one works.

![Greenshot 2020-03-17 16 59 19](https://user-images.githubusercontent.com/3807458/76875417-d2c56880-6870-11ea-8891-10d0ddbc6b33.png)
![Greenshot 2020-03-17 16 59 39](https://user-images.githubusercontent.com/3807458/76875420-d3f69580-6870-11ea-8ca7-688313eb0396.png)

**Comment by Vadorequest** [comment]
I strongly believe this behaviour is related to how `useDataProvider` proxifies the methods (such as `create`). I wasn't able to understand what's causing the behaviour though. 

This looks like a bug in the `useDataProvider` implementation.
I could use help/guidance from someone familiar with its implementation, if you don't have time to tackle the issue directly.

I'm looking for a workaround so that I can keep working on my feature, if you have any idea.

**Comment by Vadorequest** [comment]
Workaround found, doesn't crash the UI anymore.

https://github.com/UnlyEd/ra-data-graphql-prisma/pull/7
https://github.com/UnlyEd/next-right-now-admin/pull/29/commits/07b83c683b3d0a2a9a34f748d9ac7a3d8ad577cd

This doesn't fix the underlying issue.

**Comment by fzaninotto** [comment]
Hi, and thanks for your question. There seems to be a bug in your code, but as you didn't provide a link to a codesandbox, we can't see the whole context of your problem (including the data provider), so we can't really say if the problem is in your code or in react-admin. 

As explained in the [react-admin contributing guide](https://github.com/marmelab/react-admin/blob/master/.github/CONTRIBUTING.md), the right place to ask a "How To" question, get usage advice, or troubleshoot your own code, is [StackOverFlow](http://stackoverflow.com/questions/tagged/react-admin).

This makes your question easy to find by the core team, and the developer community. Unlike Github, StackOverFlow has great SEO, gamification, voting, and reputation. That's why we chose it, and decided to keep GitHub issues only for bugs and feature requests.

So I'm closing this issue, and inviting you to ask your question at:

http://stackoverflow.com/questions/tagged/react-admin

And once you get a response, please continue to hang out on the react-admin channel in StackOverflow. That way, you can help newcomers and share your expertise!

**Comment by Vadorequest** [comment]
@fzaninotto It is too complicated to reproduce for a codesandbox, I provided a OSS repository with a PR instead.

Also, it's not really related to my code, not directly. The issue resolved itself somehow but I couldn't figure out the origin. It seems to be related to the `source` value I was providing. I think changing the source was enough for this behaviour to stop.

The real issue, is how difficult it is to understand what's really happening beneath the surface, and closing this issue without going further is about the same as throwing some dirt under the carpet.

When things go wrong with react-admin, it is very hard to pinpoint the origin of the problem because of all the "magic" happening in the multiple layers, and error messages such as "cannot read data of undefined" aren't so helpful. I wish the DX would be better.

**Comment by fzaninotto** [comment]
We're doing our best, but a day only has 24 hours. 

If you want us to fix bugs, please isolate them and provide a CodeSandbox, because that's what we need to quickly spot bugs. Otherwise, navigating an unknown code base takes too long, and is a strong clue that the code is on your side. 

If you want us to improve the DX and add documentation, please post troubleshooting questions to StackOverflow instead of here, as explaines in the issue template. 

Lastly, we can take a helping hand. Try and help people on stack overflow or here so that we can focus on improving the lib.

**Comment by Vadorequest** [comment]
> If you want us to improve the DX and add documentation, please post troubleshooting questions to StackOverflow instead of here, as explaines in the issue template.

I do not understand the reasoning behind this. SO is not meant for this kind of things (DX, documentation) and it should be handled by the core team because external actors won't get it right, it needs internal knowledge of how things work and should be done to build official examples. It should be tracked through Github issues.

> If you want us to fix bugs, please isolate them and provide a CodeSandbox
It takes a lot of my own time to do so, especially when I don't know exactly how to reproduce the issue (like here) and I did my best to provide useful screenshots, explanations, live demo, OSS repository with branch to reproduce the issue, and also a video that showcases the bug. There is a limit at how much time I can spend on this, too.

I won't go further on this issue, it's too complicated to reproduce and it fixed itself somehow. 
I did my best to give actionable feedback. I guess the best I can do for the next time is to build a codesandbox with my own repository, to ease the pain of installing it yourself on your own machine.

**Comment by fzaninotto** [comment]
> I do not understand the reasoning behind this. SO is not meant for this kind of things (DX, documentation) and it should be handled by the core team because external actors won't get it right, it needs internal knowledge of how things work and should be done to build official examples. It should be tracked through Github issues.

Perhaps I wasn't clear enough, so let me rephrase. The time we spend answering troubleshooting question, we can't spend it improving the DX and documentation. That's why we chose to offload support questions to the community and why we ask people to ask for support on StackOverflow and not here on GitHub - so that we can focus, indeed, on improving the DX and documentation. And yes, GitHub issues is the right place to ask for improving DX, but it's not the right place to ask for support.

> I guess the best I can do for the next time is to build a codesandbox with my own repository, to ease the pain of installing it yourself on your own machine.

Exactly. that will be a great help, thanks in advance!
